### PR TITLE
Moving HeapSizeManager into its own class.

### PR DESCRIPTION
**Pull request by sduskis**
Did a bit of refactoring to BigtableBufferedMutator so that the its functionality is reusable.

### CODE REVIEW

**Comment by sduskis** [comment]
I should really add more testing around the executor stuff.

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+/**
+ * This class ensures that operations meet heap size and max RPC counts.  A wait will occur
+ * if RPCs are requested after heap and RPC count thresholds are exceeded.
+ */
+@VisibleForTesting class HeapSizeManager {
```
Comment (kevinsi4508) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java]: @VisibleForTesting
class

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    return operationId;
+  }
+
+  boolean isFull() {
+    return currentWriteBufferSize >= maxHeapSize
+        || pendingOperationsWithSize.size() >= maxInFlightRpcs;
+  }
+
+  @VisibleForTesting
+  synchronized void operationComplete(long operationSequenceId) {
```
Comment (kevinsi4508) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java]: nit: markOperationCompleted()

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    lastOperationChange = System.currentTimeMillis();
+    Long heapSize = pendingOperationsWithSize.remove(operationSequenceId);
+    if (heapSize != null) {
+      currentWriteBufferSize -= heapSize;
+      notifyAll();
+    } else {
+      BigtableBufferedMutator.LOG.warn("An operation completion was recieved multiple times. Your operations completed."
```
Comment (kevinsi4508) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java]: BigtableBufferedMutator.LOG ??

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java [hunk 4]
```diff
@@ -398,24 +312,20 @@ void handleExceptions() throws RetriesExhaustedWithDetailsException {
...
     public void onFailure(Throwable t) {
       addGlobalException(mutation, t);
-      sizeManager.operationComplete(operationSequenceId);
     }
 
     @Override
     public void onSuccess(GeneratedMessage ignored) {
-      sizeManager.operationComplete(operationSequenceId);
     }
   }
```
Comment (kevinsi4508) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java]: where is hasInflightRequests() called?

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        long lastUpdated = (System.currentTimeMillis() - lastOperationChange) / 1000;
+        BigtableBufferedMutator.LOG.warn("No operations completed within the last %d seconds."
+            + "There are still %d operations in progress.", lastUpdated,
+          pendingOperationsWithSize.size());
+        performedWarning = true;
+      }
+      wait(WAIT_MILLIS);
+    }
+    if (performedWarning) {
+      BigtableBufferedMutator.LOG.info("flush() completed");
```
Comment (kevinsi4508) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/HeapSizeManager.java]: ditto: change logger

File: bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestHeapSizeManager.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+/**
+ * Tests {@link HeapSizeManager}.
+ *
+ */
+public class TestHeapSizeManager {
```
Comment (kevinsi4508) [inline bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestHeapSizeManager.java]: @Junit4?

File: bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestHeapSizeManager.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+  @Mock
+  ExecutorService executorService;
```
Comment (kevinsi4508) [inline bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestHeapSizeManager.java]: nit: mockExecutorService

File: bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestHeapSizeManager.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  @Before
+  public void setup(){
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @Test
+  public void testRpcCount() throws InterruptedException{
+    HeapSizeManager underTest = new HeapSizeManager(100l, 2, executorService);
+
+    Assert.assertFalse(underTest.isFull());
```
Comment (kevinsi4508) [inline bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestHeapSizeManager.java]: static import?

File: bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java (hunk 4 above)
Comment (sduskis) [inline bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableBufferedMutator.java]: It's there for testing purposes.

**Comment by sduskis** [comment]
There's too much going on in this PR.  I'm going to create smaller PRs.

**Pull request closed without merging**